### PR TITLE
ANTLR Grammar

### PR DESCRIPTION
**Pull request by zachschoenfeld33**
### Summary
ANTLR grammar for OpenQASM3. 

### Details
- Tested by generating Python ANTLR parser
- Updated documentation
- Unit tests which compare parsed snippets of OpenQasm3 to expected parse trees
- Examples parsed and validated that no error is thrown my the grammar

### ToDo
- More unit tests (so far just have a few)
- Validate rest of examples (so far just have `adder.qasm`)

### CODE REVIEW

**Comment by ismaelfaro** [comment]
we should pass all the "examples", https://github.com/Qiskit/openqasm/tree/master/examples 

we could generate the parser for python or and do a Github action to test all of them 😁. @delapuente any suggestion on that?

File: source/grammar/qasm3.g4 [hunk 1]
```diff
@@ -1,98 +1,461 @@
...
+
+// Register Concatenation and Slicing
+concatenateExpression
+    : ASSIGN
+    ( Identifier range
+    | Identifier '||' Identifier
+    | Identifier LBRACKET expressionList RBRACKET )
+    ;
+
+range
```
Comment (ismaelfaro) [inline source/grammar/qasm3.g4:189]: it's necessary change this "range" identifier to don't collide with Python range.
- error: symbol "range" conflicts with generated code in target language or runtime

File: source/grammar/qasm3.g4 [hunk 2]
```diff
@@ -1,98 +1,461 @@
...
+LPAREN : '(' ;
+RPAREN : ')' ;
+
+COLON: ':' ;
+SEMICOLON : ';' ;
+
+DOT : '.' ;
+COMMA : ',' ;
+
+AnyString : .*? ;
```
Comment (ismaelfaro) [inline source/grammar/qasm3.g4:443]: warning: non-fragment lexer rule AnyString can match the empty string

File: source/grammar/qasm3.g4 [hunk 3]
```diff
@@ -1,98 +1,461 @@
...
+calibration
+    : calibrationGrammarDeclaration
+    | calibrationDefinition
+    ;
+
+calibrationGrammarDeclaration
+    : 'defcalgrammar' calibrationGrammar SEMICOLON
+    ;
+
+calibrationDefinition
```
Comment (ismaelfaro) [inline source/grammar/qasm3.g4:409]: warning: rule calibrationDefinition contains an optional block with at least one alternative that can match an empty string

**Review by ismaelfaro — COMMENTED** [review]
review with ANTLR Python3

**Comment by delapuente** [comment]
For testing the parser we should construct some AST and take them as reference. Sure the parser should work for the examples but I don't know if they are the best suite for testing and debugging since the ASTs of the examples are extremely complex. We could:

1. Sart by just building them and trusting the output is correct.
2. Create a series of "unit tests" for small snippets exercising the main aspects of the parser.
3. Rebuild the first set as we find problems with the current implementations.

File: source/grammar/qasm3.g4 (hunk 1 above)
Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4]: Addressed in https://github.com/Qiskit/openqasm/pull/89/commits/c3b3b28409aec991623c590a5ecfe28c7527d14c.

File: source/grammar/qasm3.g4 (hunk 2 above)
Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4]: Addressed in https://github.com/Qiskit/openqasm/pull/89/commits/c3b3b28409aec991623c590a5ecfe28c7527d14c.

File: source/grammar/qasm3.g4 (hunk 3 above)
Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4:409]: Addressed in https://github.com/Qiskit/openqasm/pull/89/commits/c3b3b28409aec991623c590a5ecfe28c7527d14c.

**Comment by zachschoenfeld33** [comment]
Have created a number of unit tests and passed `adder.qasm` example.

**Review by delapuente — DISMISSED** [review]
There is a lot of work here and we should merge it. Next steps being integrating the test in our development cycle for when changes happen in the grammar they are automatically run.

It would be also convenient to automatize the load an execution of `yaml` tests, without explicitly listing them in the test class.

However, former issues can be addressed in follow-ups. This PR presents a solid foundation on which continue working.

Nice work @zachschoenfeld33.

**Comment by zachschoenfeld33** [comment]
Incorporates https://github.com/Qiskit/openqasm/pull/122.

**Comment by zachschoenfeld33** [comment]
@taalexander @lcapelluto @delapuente @blakejohnson @steleman @mbhealy

**Comment by zachschoenfeld33** [comment]
We will also need to make sure to update this as updates are made to the spec.

**Review by delapuente — APPROVED** [review]
Moving forward, this is a good foundation. Thank you a lot @zachschoenfeld33!

**Pull request merged**